File: AKAuth0TestApp/AKAuth0TestApp/ViewController.m
//
//  ViewController.m
//  AKAuth0TestApp
//

#import "ViewController.h"
#import <Lock/Lock.h>
#import "AKChangePasswordViewController.h"

@interface ViewController ()<LoginViewControllerDelegate>

@property (copy, nonatomic) NSString *userId;

@property (weak, nonatomic) IBOutlet UILabel *usernameLabel;
@property (weak, nonatomic) IBOutlet UILabel *emailLabel;
@property (weak, nonatomic) IBOutlet UILabel *userIdLabel;

@property (weak, nonatomic) IBOutlet UITextField *emailTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;

- (IBAction)clickLoginButton:(id)sender;
- (IBAction)clickChangePasswordButton:(id)sender;

@end

@implementation ViewController

- (IBAction)clickLoginButton:(id)sender {
    if (self.emailTextField.text.length < 1) {
        [self showMessage:@"Please eneter an email"];
        return;
    }
    if (self.passwordTextField.text.length < 1) {
        [self showMessage:@"Please eneter a password"];
        return;
    }
    
    NSString *email = self.emailTextField.text;
    NSString *password = self.passwordTextField.text;
    A0APIClient *client = [[A0Lock sharedLock] apiClient];
    A0APIClientAuthenticationSuccess success = ^(A0UserProfile *profile, A0Token *token) {
        [self showUserProfile:profile];
        self.userId = profile.userId;
    };
    A0APIClientError error = ^(NSError *error){
        NSLog(@"Oops something went wrong: %@", error);
        [self clearUserProfile];
    };
    
    A0AuthParameters *params = [A0AuthParameters newDefaultParams];
    params[A0ParameterConnection] = kAuth0ConnectionType; // Or your configured DB connection
    
    //Login with email and password (Auth0 database connection)
    [client loginWithUsername:email
                     password:password
                   parameters:params
                      success:success
                      failure:error];
}

- (IBAction)clickChangePasswordButton:(id)sender {
    //Switch to Change Password ViewController
    AKChangePasswordViewController *passwordController = (AKChangePasswordViewController *)[[UIStoryboard storyboardWithName:@"Main" bundle:nil] instantiateViewControllerWithIdentifier:@"AKChangePasswordViewController"];
    passwordController.lock = [A0Lock sharedLock];
    passwordController.delegate = self;
    passwordController.email = self.emailTextField.text;
    
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:passwordController];
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {
        navController.modalPresentationStyle = UIModalPresentationFormSheet;
    }
    [self presentViewController:navController animated:YES completion:nil];
}

//Clear sessions
- (void)logout {
    A0Lock *lock = [A0Lock sharedLock];
    [[lock apiClient] logout];
    [lock clearSessions];
}

- (void)showUserProfile:(A0UserProfile *)profile {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.usernameLabel.text = profile.name;
        self.emailLabel.text = profile.email;
        self.userIdLabel.text = profile.userId;
    });
}

- (void)clearUserProfile {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.usernameLabel.text = @"";
        self.emailLabel.text = @"";
        self.userIdLabel.text = @"";
    });
}

- (void)showMessage:(NSString *)message
{
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Auth0" message:message preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
    });
}

//Delegate method
- (void)changedEmail: (NSString *)email {
    [self logout];
    [self clearUserProfile];
    self.emailTextField.text = email;
}

@end

File: AKSwiftAuth0Test/AKSwiftAuth0Test/ViewController.swift
//
//  ViewController.swift
//  AKSwiftAuth0Test
//

import UIKit
import Lock

protocol LoginViewControllerDelegate {
    func changedEmail(email:String)
}

class ViewController: UIViewController, LoginViewControllerDelegate {
    
    var userId:String?
    
    required init?(coder aDecoder: NSCoder) {
        userId = nil
        super.init(coder: aDecoder)
    }


    @IBOutlet weak var usernameLabel: UILabel!
    @IBOutlet weak var emailLabel: UILabel!
    @IBOutlet weak var userIdLabel: UILabel!
    
    @IBOutlet weak var emailTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!

    @IBAction func clickLoginButton(sender: AnyObject) {
        if (self.emailTextField.text?.characters.count < 1) {
            self.showMessage("Please eneter an email");
            return;
        }
        if (self.passwordTextField.text?.characters.count < 1) {
            self.showMessage("Please eneter a password");
            return;
        }
        
        let success = { (profile: A0UserProfile?, token: A0Token?) in
            self.userId = profile!.userId
            self.showUserProfile(profile!)
        }
        let failure = { (error: NSError) in
            self.clearUserProfile()
            print("Oops something went wrong: \(error)")
        }
        
        let email = self.emailTextField.text!;
        let password = self.passwordTextField.text!;
        let client = A0Lock.sharedLock().apiClient()
        let parameters = A0AuthParameters(dictionary: [A0ParameterConnection : kAuth0ConnectionType])
        
        //Login with email and password (Auth0 database connection)
        client.loginWithUsername(email, password: password, parameters: parameters, success: success, failure: failure)
    }

    func showUserProfile(profile: A0UserProfile) {
        dispatch_async(dispatch_get_main_queue()) {
            self.usernameLabel.text = profile.name
            self.emailLabel.text = profile.email
            self.userIdLabel.text = profile.userId
        }
    }

    func clearUserProfile() {
        dispatch_async(dispatch_get_main_queue()) {
            self.usernameLabel.text = ""
            self.emailLabel.text = ""
            self.userIdLabel.text = ""
        }
    }

    @IBAction func clickChangePasswordButton(sender: AnyObject) {
        //Switch to Change Password ViewController
        let controller:AKChangePasswordViewController = UIStoryboard.init(name:"Main", bundle:nil).instantiateViewControllerWithIdentifier("AKChangePasswordViewController") as! AKChangePasswordViewController
        controller.lock = A0Lock.sharedLock()
        controller.delegate = self
        if let actualEmail = self.emailTextField.text {
            controller.email = actualEmail
        }
        
        let navController:UINavigationController = UINavigationController.init(rootViewController: controller)
        if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiom.Pad) {
            navController.modalPresentationStyle = UIModalPresentationStyle.FormSheet
        }
        self.presentViewController(navController, animated: true, completion: nil)

    }
    
    //Clear sessions
    func logout() {
        let lock = A0Lock.sharedLock()
        lock.apiClient().logout()
        lock.clearSessions()
    }
    
    func showMessage(message: String) {
        dispatch_async(dispatch_get_main_queue()) {
            let alert = UIAlertController(title: "Auth0", message: message, preferredStyle: UIAlertControllerStyle.Alert)
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: nil))
            self.presentViewController(alert, animated: true, completion: nil)
        }
    }
    
    //Delegate method
    func changedEmail(email:String) {
        self.logout()
        self.clearUserProfile()
        self.emailTextField.text = email
    }
}
